src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs

```diff
@@ -0,0 +1,184 @@
+            var lookupResult = LookupResult.GetInstance();
+            ImmutableArray<Symbol?> withMembers;
+
+            if (withMethod is null)
+            {
+                diagnostics.Add(ErrorCode.ERR_NoSingleWithMethod, syntax.Receiver.Location, receiverType);
+                withMembers = default;
+            }
+            else
+            {
+                // Check return type
+                if (!withMethod.ReturnType.IsEqualToOrDerivedFrom(
+                    receiverType,
+                    TypeCompareKind.ConsiderEverything,
+                    ref useSiteDiagnostics))
+                {
+                    diagnostics.Add(
+                        ErrorCode.ERR_ContainingTypeMustDeriveFromWithReturnType,
+                        syntax.Receiver.Location,
+                        receiverType,
+                        withMethod.ReturnType);
+                }
+
+                // Build WithMethod member list
+                var matchingMembers = ArrayBuilder<Symbol?>.GetInstance(withMethod.ParameterCount); 
+                foreach (var p in withMethod.Parameters)
+                {
+                    this.LookupMembersInType(
+                        lookupResult,
+                        receiverType,
+                        p.Name,
+                        arity: 0,
+                        basesBeingResolved: null,
+                        options: LookupOptions.Default,
+                        originalBinder: this,
+                        diagnose: false,
+                        useSiteDiagnostics: ref useSiteDiagnostics);
+
+                    Symbol? member;
+                    if (!lookupResult.IsSingleViable)
+                    {
+                        diagnostics.Add(
+                            ErrorCode.ERR_WithParameterWithoutMatchingMember,
+                            syntax.Receiver.Location,
+                            receiverType,
+                            p.Name);
+                        member = null;
+                    }
+                    else
+                    {
+                        member = lookupResult.SingleSymbolOrDefault;
+                        Debug.Assert(member.Kind == SymbolKind.Field || member.Kind == SymbolKind.Property);
+
+                        if (!member.GetTypeOrReturnType().Equals(
+                            p.TypeWithAnnotations,
+                            TypeCompareKind.ConsiderEverything))
+                        {
+                            diagnostics.Add(
+                                ErrorCode.ERR_WithParameterTypeDoesntMatchMemberType,
+                                syntax.Receiver.Location,
+                                p.Name,
+                                p.Type,
+                                member.GetTypeOrReturnType());
+                        }
+                    }
+
+                    lookupResult.Clear();
+                    matchingMembers.Add(member);
```

Comment:
What if a member is specified twice?

Reply:
Or what if a member specified in the initializer is missing a corresponding parameter?